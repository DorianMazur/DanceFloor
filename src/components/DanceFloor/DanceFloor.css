.form-container {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    margin-bottom:20px;
}

.form-container > div{
    display: flex;
    flex-direction: row;
}

.form-container > div > input {
    margin: 0 10px;
}

.form-container > h4 {
    margin-bottom: 20px;
    width: 80%;
    max-width: 300px;
    text-align: center;
}
