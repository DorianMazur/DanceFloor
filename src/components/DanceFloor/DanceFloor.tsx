import React, { useEffect, useRef, useState } from "react";
import { Dimensions, Tile } from "../../types/canvas";
import { recreateTiles, createTiles } from "../../utils/canvasHelpers";
import Button from "../Button/Button";
import Input from "../Input/Input";
import "./DanceFloor.css";

export interface IDanceFloor {
  dimensions: Dimensions;
}

const DanceFloor: React.FC<IDanceFloor> = ({ dimensions }) => {
  const canvasRef = useRef<HTMLCanvasElement>(null);
  const generatedTiles = useRef<Tile[]>([]);
  const [columns, setColumns] = useState<null | number>(null);
  const [rows, setRows] = useState<null | number>(null);

  const generateNewTiles = (r: number, c: number) => {
    const ctx = canvasRef.current?.getContext("2d");
    if (!ctx) return;
    generatedTiles.current = createTiles(ctx, r, c);
  };

  const changeColorTile = (
    e: React.MouseEvent<HTMLCanvasElement>,
    color: string
  ) => {
    const ctx = canvasRef.current?.getContext("2d");
    if (!ctx || !canvasRef.current) return;
    const x = e.clientX - canvasRef.current?.offsetLeft;
    const y = e.clientY - canvasRef.current?.offsetTop;
    recreateTiles(ctx, x, y, generatedTiles.current, color);
  };

  useEffect(() => {
    generateNewTiles(dimensions.rows, dimensions.columns);
  }, [dimensions]);

  return (
    <>
      <div className="form-container">
        <h4>
          If you do not specify the configuration, the default values ​​taken
          from the API will be used
        </h4>
        <div>
          <Input
            label="Set column quantity"
            value={columns}
            onChange={(v) => setColumns(v)}
            name="columns"
          />
          <Input
            label="Set row quantity"
            value={rows}
            onChange={(v) => setRows(v)}
            name="rows"
          />
          <Button
            text="Generate"
            onPress={() =>
              generateNewTiles(
                rows || dimensions.rows,
                columns || dimensions.columns
              )
            }
          />
        </div>
      </div>
      <canvas
        className="canvas"
        ref={canvasRef}
        width="500"
        onMouseMove={(e) => changeColorTile(e, "rgb(0,0,0)")}
        onClick={(e) => changeColorTile(e, "rgb(255,255,255)")}
      />
    </>
  );
};

export default DanceFloor;
