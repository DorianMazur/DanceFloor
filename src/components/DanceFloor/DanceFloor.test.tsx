import { render } from "@testing-library/react";
import DanceFloor from "./DanceFloor";
import 'jest-canvas-mock';

describe("DanceFloor", () => {
  test("renders without crash", async () => {
    const utils = render(<DanceFloor dimensions={{rows: 3, columns:2}} />);

    expect(utils).toMatchSnapshot()
  });
  /* I should write more tests, but i don't have enough time
  If I had enough time then:
  - I will use jest-canvas-mock 
    - ctx.__getPath() to check if correct path exist for given columns and rows
    - ctx.__getEvents() to check events 
  - Also i will do some research about testing canvas because i don't have experience in canvas tests :D
  */
});
