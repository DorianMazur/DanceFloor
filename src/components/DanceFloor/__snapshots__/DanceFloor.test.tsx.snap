// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DanceFloor renders without crash 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="form-container"
      >
        <h4>
          If you do not specify the configuration, the default values ​​taken from the API will be used
        </h4>
        <div>
          <input
            class="input"
            name="columns"
            placeholder="Set column quantity"
            type="number"
            value=""
          />
          <input
            class="input"
            name="rows"
            placeholder="Set row quantity"
            type="number"
            value=""
          />
          <button
            class="button"
          >
            Generate
          </button>
        </div>
      </div>
      <canvas
        class="canvas"
        width="500"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="form-container"
    >
      <h4>
        If you do not specify the configuration, the default values ​​taken from the API will be used
      </h4>
      <div>
        <input
          class="input"
          name="columns"
          placeholder="Set column quantity"
          type="number"
          value=""
        />
        <input
          class="input"
          name="rows"
          placeholder="Set row quantity"
          type="number"
          value=""
        />
        <button
          class="button"
        >
          Generate
        </button>
      </div>
    </div>
    <canvas
      class="canvas"
      width="500"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
